src/v/kafka/server/handlers/fetch.cc

```diff
@@ -1282,32 +1282,51 @@ size_t op_context::fetch_partition_count() const {
     }
 }
 
+namespace {
+
+// Calls f for each topic in the request, passing the associated partitions.
+// This exists to adapt the sessionfull and sessionless fetch cases to a
+// consistent iteration interface.
 template<typename Func>
-void op_context::for_each_fetch_partition(Func&& f) const {
+void for_each_fetch_partition(const op_context& octx, const Func& f) {
     /**
      * Iterate over original request only if it is sessionless or initial
      * full fetch request. For not initial full fetch requests we may
      * leverage the fetch session stored partitions as session was populated
      * during initial pass. Using session stored partitions will account for
      * the partitions already read and move to the end of iteration order
      */
+    auto& session_ctx = octx.session_ctx;
+
     if (
       session_ctx.is_sessionless()
-      || (session_ctx.is_full_fetch() && initial_fetch)) {
```

Comment:
Here we do the "hierachy" change. Instead of calling the callback with flat list of `fetch_session_partition`, we call the callback with a topic name + range of partitions. The type of range is different in the sessionful and session-less cases, but in both cases they yield the same element type.